File: web/src/main/java/zuk/sast/rules/controller/stock/AllStockController.java
package zuk.sast.rules.controller.stock;

import com.beust.jcommander.internal.Lists;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import zuk.sast.rules.controller.mapper.StockMapper;
import zuk.sast.rules.controller.mapper.entity.StockEntity;

import java.util.*;
import java.util.stream.Collectors;

import zuk.sast.rules.controller.stock.analysis.LoaderStockData;

/***
 *
 */
@Slf4j
@RestController
@RequestMapping("stock")
public class AllStockController {

    //https://xtrade.newone.com.cn/market/json?funcno=30029&version=1&stock_code=300287&market=SZ&start=0&dayitems=1&cms-trace-id=fe98fa982fbf42ce88b268b841bd8a84.1756699944822.16897853520034
    //https://xtrade.newone.com.cn/market/json?funcno=20006&version=1&stock_code=300287&market=SZ&count=100&cms-trace-id=fe98fa982fbf42ce88b268b841bd8a84.1756699944822.16897853520033
    //https://xtrade.newone.com.cn/market/json?funcno=30029&version=1&stock_code=300287&market=SZ&start=0&dayitems=5&cms-trace-id=fe98fa982fbf42ce88b268b841bd8a84.1756700218269.16897853520262
    //https://xtrade.newone.com.cn/market/json?funcno=20044&version=1&stock_code=300287&market=SZ&count=1000&type=day&cms-trace-id=fe98fa982fbf42ce88b268b841bd8a84.1756700327368.16897853520352

    @Autowired
    private StockMapper stockMapper;

    @GetMapping("delete")
    public synchronized Map<String, Object> delete(String api_code) {
        log.info("delete:" + api_code);

        Map<String, Object> result = new HashMap<>();
        try {
            this.stockMapper.deleteByCode(api_code);

            result.put("status", "ok");
            return result;

        }catch (Exception e) {
            //e.printStackTrace();
            log.error(e.getMessage());
            result.put("status", e.getMessage());
        }

        return result;
    }


    @GetMapping("add")
    public synchronized Map<String, Object> add(String api_code) {
        log.info("add:" + api_code);
        Map<String, Object> result = new HashMap<>();
        try {
            List list = this.stockMapper.selectByCode(api_code);
            if(list!=null && list.size()>0){
                result.put("status", "已存在");
                return result;
            }
            else {
                LoaderStockData.STOCKS.stream().filter(stock->stock.getApi_code().equals(api_code)).forEach(socket->{
                    StockEntity stockEntity = new StockEntity();
                    stockEntity.setId(UUID.randomUUID().toString());
                    stockEntity.setCode(socket.getApi_code());
                    stockEntity.setJys(socket.getJys());
                    stockEntity.setName(socket.getName());
                    stockEntity.setType("LOVE");
                    this.stockMapper.insert(stockEntity);
                });
                result.put("status", "ok");
                return result;
            }

        }catch (Exception e) {
            //e.printStackTrace();
            log.error(e.getMessage());
            result.put("status", e.getMessage());
        }

        return result;
    }

    /***
     * 我的关注
     * @return
     */
    @GetMapping("my")
    public Map<String, Object> my(){
        try {
            List<StockEntity> stockEntities = stockMapper.selectAll();

            Map<String, Object> map = new HashMap<>();
            Set<String> sets = stockEntities.stream().map(e->e.getCode()).collect(Collectors.toSet());
            List<LoaderStockData.StockApiVO> ls = LoaderStockData.STOCKS.stream().filter(stock->{
                return sets.contains(stock.getApi_code());
            }).toList();

            map.put("stocks", ls);
            return map;
        }
        catch (Exception e) {
            e.printStackTrace();
            log.error(e.getMessage());
        }
        return new HashMap<>();
    }

    public static final Set<String> tags = Arrays.asList(
                    "人工智能",
                    "deepseek",
                    "昇腾",
                    "数据中心",
                    "芯片"
            ).stream().collect(Collectors.toSet());

    @GetMapping("all")
    public Map<String, Object> all(String search){

        List<LoaderStockData.StockApiVO> list;
        if(search!=null && search.length()>0){
            Set<String> splits = Arrays.stream(search.split("\n")).filter(e->e!=null && e.trim().length()>0).collect(Collectors.toSet());

            list = LoaderStockData.STOCKS.stream().filter(e->{
                return splits.stream().filter(tag->{
                    return e.getGl().contains(tag);
                }).count() > 0;
            }).collect(Collectors.toList());

        }
        else {
            //没有输入条件，则默认输出1000条
            list = LoaderStockData.STOCKS.stream().filter(e->{
                return tags.stream().filter(tag->{
                    return e.getGl().contains(tag);
                }).count() > 0;
            }).collect(Collectors.toList());
            if(list!=null && list.size()>1000){
                list = list.subList(0, 1000);
            }
        }

        log.info("search:" + search +  ", stocks:" + list.size() + ", blocks:" + tags.size());

        Map<String, Object> map = new HashMap<>();
        map.put("stocks", list);
        map.put("blocks",tags);
        return map;
    }

}

File: web/stockapi/src/main/scala/zuk/sast/controller/AllStockController.scala
package zuk.sast.controller

import org.apache.commons.csv.CSVFormat
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.{GetMapping, RequestMapping, RestController}
import zuk.sast.controller.mapper.StockMapper
import zuk.sast.controller.mapper.entity.StockEntity
import zuk.stockapi.{LoaderLocalStockData, StockApiVo}
import zuk.tu_share.dto.HmDetail

import java.util
import java.util.stream.Collectors
import java.util.{Arrays, HashMap, List, Map, Set, UUID}
import java.io.{File, FileReader}
import java.nio.charset.Charset
import scala.jdk.CollectionConverters.*

object AllStockController{
  val hmDetailMap = scala.collection.mutable.HashMap[String, List[HmDetail]]()
}

@RestController
@RequestMapping(value = Array("stock"))
class AllStockController {

  private val log = LoggerFactory.getLogger(classOf[AllStockController])

  @Autowired
  private var stockMapper: StockMapper = _

  @GetMapping(value = Array("delete"))
  def delete(api_code: String): util.Map[String, Object] = {
    log.info("delete:" + api_code)
    val result: util.Map[String, AnyRef] = new util.HashMap[String, AnyRef]
    try {
      this.stockMapper.deleteByCode(api_code)
      result.put("status", "ok")
      return result
    } catch {
      case e: Exception =>
        //e.printStackTrace();
        log.error(e.getMessage)
        result.put("status", e.getMessage)
    }
    result
  }

  @GetMapping(value = Array("add"))
  def add(api_code: String): util.Map[String, Object] = {
    log.info("add:" + api_code)
    val result = new util.HashMap[String, Object]
    try {
      val list = this.stockMapper.selectByCode(api_code)
      if (list != null && list.size > 0) {
        result.put("status", "已存在")
        return result
      }
      else {
        LoaderLocalStockData.STOCKS.stream.filter((stock: StockApiVo) => stock.getApi_code == api_code).forEach((socket: StockApiVo) => {
          val stockEntity = new StockEntity
          stockEntity.id = UUID.randomUUID.toString
          stockEntity.code = socket.getApi_code
          stockEntity.jys = socket.getJys
          stockEntity.name = socket.getName
          this.stockMapper.insert(stockEntity)

        })
        result.put("status", "ok")
        return result
      }
    } catch {
      case e: Exception =>
        //e.printStackTrace();
        log.error(e.getMessage)
        result.put("status", e.getMessage)
    }
    result
  }

  /** *
   * 我的关注
   *
   * @return
   */
  @GetMapping(value = Array("my")) def my: util.Map[String, Object] = {
    try {
      val stockEntities = stockMapper.selectAll()
      val map = new util.HashMap[String, AnyRef]
      val sets = stockEntities.stream.map((e: StockEntity) => e.code).collect(Collectors.toSet)
      val ls = LoaderLocalStockData.STOCKS.stream.filter((stock: StockApiVo) => {
        sets.contains(stock.getApi_code)

      }).toList
      map.put("stocks", ls)
      return map
    } catch {
      case e: Exception =>
        e.printStackTrace()
        log.error(e.getMessage)
    }
    new util.HashMap[String, Object]
  }

  /***
   * 游资交易每日明细
   */
  @GetMapping(value = Array("all"))
  def all(tradedate: String, search: String): Map[String, Object] = {
    val hmFile = new File(s"tushare/hm/hm_detail/hm_detail-${tradedate}.csv") //龙虎榜路径
    println(s"游资交易每日明细:hmPath=${hmFile.exists()}, tradedate=${tradedate}, search=${search}")
    if(hmFile.exists() && AllStockController.hmDetailMap.get(tradedate).isEmpty){
      //路径存在
      val in = new FileReader(hmFile.getAbsolutePath, Charset.forName("UTF-8"))
      val records = CSVFormat.DEFAULT.withFirstRecordAsHeader().parse(in)
      val codes = records.asScala.map(record => {
          val hmDetail = new HmDetail()
          hmDetail.trade_date = record.get("trade_date")
          hmDetail.ts_code = record.get("ts_code")
          hmDetail.ts_name = record.get("ts_name")
          hmDetail.buy_amount = record.get("buy_amount")
          hmDetail.sell_amount = record.get("sell_amount")
          hmDetail.net_amount = record.get("net_amount")
          hmDetail.hm_name = record.get("hm_name")
          hmDetail.hm_orgs = record.get("hm_orgs")
          //
          hmDetail.splitTsCode(hmDetail.ts_code)
          hmDetail
        })
        .toList
      in.close()
      val countMap = codes.groupBy(_.ts_code).map(e=>(e._1, e._2.size))
      codes.foreach(c=>{
        c.count = countMap.get(c.ts_code).get //计算买入的游资数量
      })
      AllStockController.hmDetailMap.put(tradedate, codes.sortBy(_.count).reverse.asJava)
    }
    val list = new util.ArrayList[HmDetail]()
    if(AllStockController.hmDetailMap.get(tradedate).nonEmpty){
      list.addAll(AllStockController.hmDetailMap.get(tradedate).get.asScala.filter(e=>{
          scala.collection.mutable.ListBuffer(e.ts_code, e.ts_name, e.hm_name, e.hm_orgs).filter(_.contains(search)).size>0
      }).asJava)
    }
    val map: Map[String, AnyRef] = new HashMap[String, AnyRef]
    map.put("stocks", list)
    map
  }

}
